src/trace/trace.c

```diff
@@ -96,18 +98,127 @@ static inline void mtrace_event(const char *data, uint32_t length)
 	platform_shared_commit(trace, sizeof(*trace));
 }
 
+#if CONFIG_TRACE_FILTERING_VERBOSITY
 /**
- * \brief Runtime trace filtering
+ * \brief Runtime trace filtering based on verbosity level
  * \param lvl log level (LOG_LEVEL_ ERROR, INFO, DEBUG ...)
  * \param uuid uuid address
  * \return false when trace is filtered out, otherwise true
  */
-static inline bool trace_filter_pass(uint32_t lvl,
-				     const struct tr_ctx *ctx)
+static inline bool trace_filter_verbosity(uint32_t lvl, const struct tr_ctx *ctx)
 {
 	/* LOG_LEVEL_CRITICAL has low value, LOG_LEVEL_VERBOSE high */
 	return lvl <= ctx->level;
 }
+#endif /* CONFIG_TRACE_FILTERING_VERBOSITY */
+
+#if CONFIG_TRACE_FILTERING_ADAPTIVE
+struct recent_log_entry {
+	uint32_t entry_id;
+	uint64_t message_ts;
+#if CONFIG_TRACE_RECENT_PROLONG
+	uint64_t first_suppression_ts;
+#endif /* CONFIG_TRACE_RECENT_PROLONG */
+	uint32_t suppress_count;
+};
+
+static struct recent_log_entry recent_entries[CONFIG_TRACE_RECENT_ENTRIES_COUNT];
+static struct recent_log_entry entry_to_emit;
+
+static inline void prepare_to_emit(struct recent_log_entry *entry)
+{
+	if (entry->suppress_count)
+		memcpy_s(&entry_to_emit, sizeof(entry_to_emit), entry, sizeof(*entry));
+	memset(entry, 0, sizeof(*entry));
+}
+
+static inline void emit_recent_entry(void)
+{
+	if (!entry_to_emit.suppress_count)
+		return;
+
+	_log_message(trace_log_unfiltered, false, LOG_LEVEL_INFO, _TRACE_INV_CLASS, &dt_tr,
+		     _TRACE_INV_ID, _TRACE_INV_ID, "Suppressed %u similar messages: %s",
+		     entry_to_emit.suppress_count, entry_to_emit.entry_id);
+
+	memset(&entry_to_emit, 0, sizeof(entry_to_emit));
+}
+
+static inline void emit_recent_entries(uint64_t current_ts)
+{
+	int i;
+
+	if (entry_to_emit.entry_id)
+		emit_recent_entry();
+
+	for (i = 0; i < CONFIG_TRACE_RECENT_ENTRIES_COUNT; i++) {
+		if (current_ts - recent_entries[i].message_ts >
+			CONFIG_TRACE_RECENT_TIME_THRESHOLD) {
+			prepare_to_emit(&recent_entries[i]);
+			emit_recent_entry();
+		}
+	}
+}
+
+/**
+ * \brief Runtime trace flood suppresion
+ * \param log_level log verbosity level
+ * \param entry ID of log entry
+ * \param message_ts message timestamp
+ * \return false when trace is filtered out, true otherwise
+ */
+static inline bool trace_filter_flood(uint32_t log_level, uint32_t entry, uint64_t message_ts)
+{
+	struct recent_log_entry *oldest_entry = recent_entries;
+	int i;
+
+	/* don't attempt to suppress debug messages using this method, it would be uneffective */
+	if (log_level == LOG_LEVEL_DEBUG)
```

Comment:
We should probably say <= or >= here to future proof more verbose debug message levels.

Reply:
ping

Reply:
done